File: MTGSearch/src/main/java/com/dbottillo/mtgsearchfree/model/DeckBucket.java
package com.dbottillo.mtgsearchfree.model;

import com.dbottillo.mtgsearchfree.util.LOG;

import java.util.ArrayList;
import java.util.List;

public class DeckBucket extends CardsBucket {

    private ArrayList<MTGCard> creatures = new ArrayList<>();
    private ArrayList<MTGCard> instantAndSorceries = new ArrayList<>();
    private ArrayList<MTGCard> other = new ArrayList<>();
    private ArrayList<MTGCard> lands = new ArrayList<>();
    private ArrayList<MTGCard> side = new ArrayList<>();


    public ArrayList<MTGCard> getCreatures() {
        return creatures;
    }

    public ArrayList<MTGCard> getInstantAndSorceries() {
        return instantAndSorceries;
    }

    public ArrayList<MTGCard> getOther() {
        return other;
    }

    public ArrayList<MTGCard> getLands() {
        return lands;
    }

    public ArrayList<MTGCard> getSide() {
        return side;
    }

    @Override
    public void setCards(List<MTGCard> cards) {
        for (MTGCard card : cards) {
            if (card.isSideboard()) {
                side.add(card);
            } else if (card.isLand()) {
                lands.add(card);
            } else if (card.getTypes().contains("Creature")) {
                creatures.add(card);
            } else if (card.getTypes().contains("Instant") || card.getTypes().contains("Sorcery")) {
                instantAndSorceries.add(card);
            } else {
                other.add(card);
            }
        }
    }

    @Override
    public List<MTGCard> getCards() {
        LOG.d("deck of bucket " + key + " requested");
        ArrayList<MTGCard> cards = new ArrayList<>(creatures);
        cards.addAll(instantAndSorceries);
        cards.addAll(other);
        cards.addAll(lands);
        cards.addAll(side);
        return cards;
    }

    @Override
    public String toString() {
        return "DeckBucket: [nCreatures:" + creatures.size() + ", nLands" + lands.size()
                + ", nInstantAndSorceries:" + instantAndSorceries.size() + ", nOther:" + other.size() + "]";
    }

    public int numberOfCards() {
        List<MTGCard> cards = getCards();
        int total = 0;
        for (MTGCard card : cards) {
            total += card.getQuantity();
        }
        return total;
    }

    public int numberOfUniqueCards() {
        return creatures.size() + instantAndSorceries.size() + other.size() + lands.size() + side.size();
    }

    public int numberOfCardsWithoutSideboard() {
        return numberOfCards() - numberOfCardsInSideboard();
    }

    public int numberOfCardsInSideboard() {
        return totalQuantityOfList(side);
    }

    public int numberOfUniqueCardsInSideboard() {
        return side.size();
    }

    public int getNumberOfUniqueCreatures() {
        return creatures.size();
    }

    public int getNumberOfCreatures() {
        return totalQuantityOfList(creatures);
    }

    public int getNumberOfUniqueInstantAndSorceries() {
        return instantAndSorceries.size();
    }

    public int getNumberOfInstantAndSorceries() {
        return totalQuantityOfList(instantAndSorceries);
    }

    public int getNumberOfUniqueLands() {
        return lands.size();
    }

    public int getNumberOfLands() {
        return totalQuantityOfList(lands);
    }

    public int getNumberOfUniqueOther() {
        return other.size();
    }

    public int getNumberOfOther() {
        return totalQuantityOfList(other);
    }

    private static int totalQuantityOfList(List<MTGCard> list){
        int total = 0;
        for (MTGCard card : list) {
            total += card.getQuantity();
        }
        return total;
    }
}

File: MTGSearch/src/main/kotlin/com/dbottillo/mtgsearchfree/model/DeckBucket.kt
package com.dbottillo.mtgsearchfree.model

data class DeckBucket(
    val creatures: MutableList<MTGCard> = mutableListOf(),
    val instantAndSorceries: MutableList<MTGCard> = mutableListOf(),
    val other: MutableList<MTGCard> = mutableListOf(),
    val lands: MutableList<MTGCard> = mutableListOf(),
    val side: MutableList<MTGCard> = mutableListOf()
) {

    var cards: List<MTGCard>
        get() {
            return creatures + instantAndSorceries + other + lands + side
        }
        set(cards) {
            for (card in cards) {
                if (card.isSideboard) {
                    side.add(card)
                } else if (card.isLand) {
                    lands.add(card)
                } else if (card.types.contains("Creature")) {
                    creatures.add(card)
                } else if (card.types.contains("Instant") || card.types.contains("Sorcery")) {
                    instantAndSorceries.add(card)
                } else {
                    other.add(card)
                }
            }
        }

    val numberOfUniqueCreatures: Int
        get() = creatures.size

    val numberOfCreatures: Int
        get() = totalQuantityOfList(creatures)

    val numberOfUniqueInstantAndSorceries: Int
        get() = instantAndSorceries.size

    val numberOfInstantAndSorceries: Int
        get() = totalQuantityOfList(instantAndSorceries)

    val numberOfUniqueLands: Int
        get() = lands.size

    val numberOfLands: Int
        get() = totalQuantityOfList(lands)

    val numberOfUniqueOther: Int
        get() = other.size

    val numberOfOther: Int
        get() = totalQuantityOfList(other)

    fun numberOfCards(): Int {
        return totalQuantityOfList(cards)
    }

    fun numberOfUniqueCards(): Int {
        return creatures.size + instantAndSorceries.size + other.size + lands.size + side.size
    }

    fun numberOfCardsWithoutSideboard(): Int {
        return numberOfCards() - numberOfCardsInSideboard()
    }

    fun numberOfCardsInSideboard(): Int {
        return totalQuantityOfList(side)
    }

    fun numberOfUniqueCardsInSideboard(): Int {
        return side.size
    }

    private fun totalQuantityOfList(list: List<MTGCard>): Int {
        return list.sumBy { it.quantity }
    }
}
